src/main/java/proguard/classfile/editor/ClassReferenceFixer.java

```diff
@@ -858,6 +859,22 @@ public static String shortKotlinNestedClassName(String enclosingClassName,
         }
     }
 
+
+    /**
+     * Returns a demangled Kotlin name.
+     *
+     * Mangled names are generated by the Kotlin compiler to avoid name clashes;
+     * these names are the same as the original names, but with a suffix appended.
+     *
+     * The function name in the Kotlin metadata should use the demangled name.
+     */
+    private static String demangle(String name)
+    {
+        int endIndex = name.indexOf('-');
+        return endIndex == -1 ? name : name.substring(0, endIndex);
+    }
```

Comment:
Not sure, but will it break in case of combining backticks with the dash symbol?
```Kotlin
fun `useCase-123`(): Result<Int>
```

Reply:
Indeed, that would be a problem! The demangling function will have to be smarter to avoid that. Unfortunately, there's no `isMangled` flag to easily tell if it's mangled or just has `-`. The mangling function is here https://github.com/JetBrains/kotlin/blob/master/compiler/backend/src/org/jetbrains/kotlin/codegen/state/inlineClassManglingUtils.kt

Reply:
I have a slightly broader question, if I may. What's the purpose of the "update metadata" step? It is given as a fact in the documentation without further explanation, but I don't understand, which problem does it solve?

Reply:
Do you mean `-keepkotlinmetadata` in general, or specifically the `ClassReferenceFixer`?

The general updating of metadata step is to keep the metadata in-sync with the obfuscated JVM names. Before ProGuard added support for Kotlin metadata (<7.0.0) the metadata was left untouched. This meant that metadata remained attached but contained unobfuscated names that did not match the obfuscated JVM names. When using `-keepkotlinmetadata` ProGuard obfuscates JVM names as usual, but then those name changes are reflected in the attached metadata. 

Many parts of the metadata have an underlying JVM component, for example, functions in the metadata don't contain code but they link to an underlying JVM method. So, when syncing the names, the name of the function in the metadata is updated to that of the underlying JVM method.

The `ClassReferenceFixer` is a general utility that is used in many places in ProGuard to "fix" names according to their new name (for Java names as well as Kotlin names). In this case, the function name had not actually change but the underlying JVM name did not match originally anyway, so the name gets updated to the underlying JVM method name incorrectly.

Reply:
Ok, I see, thank you for the explanation. What I mean, is that according to the https://www.guardsquare.com/manual/languages/kotlin
> ProGuard will only keep the Kotlin metadata of a class if you explicitly keep that class or one of its members and you add `-keepkotlinmetadata` option to your configuration.

So what's the purpose of "fixing" the names in the metadata, if the underlying JVM names are not obfuscated (`keep` is a pre-requirement) in the first place?

Reply:
This is more due to the architecture of ProGuard(CORE) - the `ClassReferenceFixer` has no knowledge of any renaming step (which could also happen in places other than obfuscation).

The `ClassReferenceFixer` has access only to the current name of something (e.g. a Kotlin function) and the name of the referenced entity (e.g. a JVM method). The assumption for Kotlin metadata has been that these names should be in-sync (which doesn't hold true for mangled names). And there is, for example, no flag on the referenced JVM method that informs the `ClassReferenceFixer` that it has changed. The convention in non-Kotlin part of `ClassReferenceFixer` is to check the old name against the new name, and update if it differs.

Reply:
I see, thank you for the explanation!